### ISSUE TITLE
SO_REUSEPORT crashes with Invalid argument.

### ISSUE THREAD

**Issue by jonathanslenders** [eventlet/eventlet#380]
```python
        if sys.platform[:3] != "win":
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        if hasattr(socket, 'SO_REUSEPORT'):
            # NOTE(zhengwei): linux kernel >= 3.9
>           sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)

E           OSError: [Errno 22] Invalid argument
```

Tested on the Linux subsystem for Windows, but it also appears the case in our Docker CI environment.

**Comment by temoto** [comment]
Please, clarify this:
you have a system where socket.SO_REUSEPORT is defined, but trying to use
it raises an exception?

**Comment by jonathanslenders** [comment]
That is right, yes. No idea how that is possible. Catching OSError around ``sockt.setsockopt`` seems to handle it.

**Comment by temoto** [comment]
That's clearly bug in Python or in that Windows/Linux poor child.

On Jan 24, 2017 4:25 PM, "Jonathan Slenders" <notifications@github.com>
wrote:

That is right, yes. No idea how that is possible. Catching OSError around
sockt.setsockopt seems to handle it.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/eventlet/eventlet/issues/380#issuecomment-274801673>,
or mute the thread

**Comment by temoto** [comment]
I'm glad to help but we need a better idea than catching OSError and hiding (possibly valuable) errors.

**Comment by Dakkaron** [comment]
This has been addressed by over at the Python bugtracker. The same problem also exists on Android, by the looks of it. Over there they said, you should just catch the exception.

https://github.com/python/asyncio/pull/418

Their solution was to catch the OSError and raise a ValueError instead. This is not yet in Python 3.6.1, so you would probably have to catch both OSError and ValueError.

Catching these errors on this location should not be able to hide any valuable errors, since that operation is optional anyways. It does help on systems where it is supported and it does nothing on systems where it is not supported. The try/catch can be implemented just around this one function call, so there is nothing else that can be hidden by this.

**Comment by temoto** [comment]
Please, try this patch.

`pip install https://github.com/eventlet/eventlet/archive/664f254c7f9953954925c4f8439a79171e4dc62d.zip`

**Comment by temoto** [comment]
And just to make it clear:

* I condole to guys having to use Windows-Linux mix (run while you can)
* the "official" fix in python/asyncio#418 of just ignoring any OSError without judgement is creating more trouble in future, as any hiding of errors always was. Unfortunately, we are likely to see bug reports in random server software such as "spent six hours digging all libraries to find asyncio ignored an important error".

This patch checks `OSError.errno` at least, which does not pinpoint SO_REUSEPORT availability, but it's better than nothing.

**Comment by Dakkaron** [comment]
Thanks for the patch, it seems to work fine. There are two little things that I would maybe change though:

First, it says "On Windows: known bug, wontfix". I would change this to "On WSL" or "On Linux on Windows", since pure Windows Python does not have that problem.

Second, I would probably also catch ValueError, since the official fix does not just ignore the `OSError`, but reraises it as a `ValueError`. Here is the relevant part of the code:

```if reuse_port:
	if not hasattr(socket, 'SO_REUSEPORT'):
		raise ValueError(
			'reuse_port not supported by socket module')
	else:
		try:
			sock.setsockopt(
				socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
		except OSError:
			raise ValueError((
				'reuse_port not supported by socket module, '
				'SO_REUSEPORT defined but not implemented.'))```

I am with you that this fix will cause more trouble, since now to be compatible, you basically have to catch both `OSError` (to support current and older versions of Python) and `ValueError` (to support future versions of Python).

The problem is actually not really Python's fault. From what I understand `SO_REUSEPORT` is defined at compile time. And WSL runs unmodified Linux programs on a compatibility layer on Windows. This means, at compile time the compiler actually believes that the target system can use `SO_REUSEPORT` and there is hardly any real way to catch the problem there.
Unfortunately, Windows is lacking that feature (which is why `SO_REUSEPORT` is not defined on pure Windows Python) and thus WSL cannot implement it.
Similar things seem to happen with Docker and Android, from what I have read.

This is surprisingly the only problem I have had with WSL so far and I have used it quite extensively over the last few weeks. It causes by far less problems than Cygwin or trying to run all the stuff I need (Python and some other tools) natively on Windows.

**Comment by temoto** [comment]
@Dakkaron AFAIK they patched asyncio library which is not used by Eventlet, so ValueError is irrelevant?

Compile, link, run -time - mother always sees her child as innocent and finds words to his fair, but the truth is software is connected at strictly defined APIs. Presenting platform features either is Python's responsibility or it's not, no matter how it was compiled. If they say applications must probe platform features, that's more job for us but it's actually great news - clear who should do what. Currently it skips some checks at certain circumstances, bla-bla - complexity is the enemy.

**Comment by Dakkaron** [comment]
I thought asyncio is responsible for the socket functions. I guess I was wrong. Nevermind then, sorry for that.

I understand what you mean. I think that the documentation of Python's standard library should be a lot more open with what exceptions the functions can raise. I ran into troubles with that quite a few times by now that I thought that I caught all relevant exceptions only to discover that I didn't catch some obscure exception that only occurred at some instances.

Thanks for fixing that problem, it helps a lot!

**Comment by temoto** [comment]
Thanks for testing.

Fix is merged in master 09e166bfbd6ba6f52477a7ae1c421dcabd1f004c and will be available in next PyPI release.

**Comment by jonathanslenders** [comment]
Thank you, everyone!

**Comment by JacoFourie** [comment]
Hi all. I am using Flask_socketio and it uses Eventlet. All is working on Raspbian on the Raspberry Pi. I now am playing around with an Orange Pi Zero using Armbian as OS. Everything works acept Eventlet. I get the Protocol is not suported error. What can I do to fix it ?

**Comment by JacoFourie** [comment]
OK so if I comment out the line "sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)" then my app starts. What wont work if I do that ?

**Comment by temoto** [comment]
@JacoFourie hello.

As you can see above, the fix that turns this error into warning (which you can set to ignore), so please update eventlet and all good.

`SO_REUSEPORT` enables multiple independent processes to bind to same address. OS decides which process accepts next connection.

**Comment by fzzzy** [comment]
@JacoFourie I think from reading the comments on this bug, your problem is a bug in the python you are using, not eventlet.

If you comment out the SO_REUSEPORT line, you may get an EADDRINUSE error if a server process which had previously bound to the same port is still shutting down and the old socket is in the TIME_WAIT state. This means you should just wait at least TIME_WAIT seconds after shutting down the old server, before starting up the new server.

**Comment by temoto** [comment]
@fzzzy sorry mate, I believe you're describing `SO_REUSEADDR`.

**Comment by JacoFourie** [comment]
@temoto What should I do to have it send a warning. I did copy the new file in place but I still got the error.  Do I add some code to my Python scrip or to the library file ?

**Comment by fzzzy** [comment]
@temoto You're probably right. What effect would commenting out SO_REUSEPORT have, then?

**Comment by fzzzy** [comment]
@JacoFourie Update to the latest version of eventlet.

**Comment by JacoFourie** [comment]
I did. I still get this error.

`Starting webapp...
Traceback (most recent call last):
  File "./web_main.py", line 177, in <module>
    socketio.run(app, host='0.0.0.0', port=23432, debug=False)
  File "/usr/local/lib/python2.7/dist-packages/flask_socketio/__init__.py", line 511, in run
    run_server()
  File "/usr/local/lib/python2.7/dist-packages/flask_socketio/__init__.py", line 490, in run_server
    eventlet_socket = eventlet.listen(addresses[0][4], addresses[0][0])
  File "/usr/local/lib/python2.7/dist-packages/eventlet/convenience.py", line 64, in listen
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
  File "/usr/lib/python2.7/socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 92] Protocol not available
`

**Comment by JacoFourie** [comment]
I installed eventlet using

`pip install -U https://github.com/eventlet/eventlet/archive/master.zip`

and also just copied the file. I still get the same error. Only if I comment that line out then it works.

**Comment by temoto** [comment]
@fzzzy effect: you run one program and it binds to addres, you run same program (without fork-child relationship) and it gets erro "address already in use". With `SO_REUSEPORT` many processes (with same effective UID) can bind.

**Comment by temoto** [comment]
I didn't know that `socket.error` has become a subclass of `OSError` only in Python 3.

Please, try again with this patch.

`pip install https://github.com/eventlet/eventlet/archive/f1b63abd6db186c978077499f9670600da599d1a.zip`

**Comment by temoto** [comment]
@JacoFourie I didn't know that `socket.error` has become a subclass of `OSError` only in Python 3.

Please, try again with this patch.

`pip install https://github.com/eventlet/eventlet/archive/f1b63abd6db186c978077499f9670600da599d1a.zip`

**Comment by JacoFourie** [comment]
OK cool now I get only the warning. Thanks for the help. You are a star !!!

**Comment by temoto** [comment]
Thanks for testing, Python2 fix is merged in master f1b63abd6db186c978077499f9670600da599d1a

> So all that wont work is I can not start another app on the same port ?

Another process can't bind to same address:port pair. It may be same app or another. Bind-fork scenario still works, of course.
